src/library/Cardano/Fee.hs

```diff
@@ -331,26 +347,84 @@ coverRemainingFee (Fee fee) = go [] where
+                      & fmap payFee
+                      & coalesceDust (dustThreshold opts)
+            reduceChangeOutputs opts (s { change = chgs' })
+
+        -- the selection is leaving more fee than necessary. This typically
+        -- occurs when, after reducing a an output to pay for existing fees,
+        -- fees turn out to be less expensive than originally. In which, the
+        -- behavior will depend on whether or not the node allows transaction to
+        -- be unbalanced or not.
+        Just δ_original | δ_original > φ_original -> do
+            let Just extraChng = δ_original `C.sub` φ_original -- cf above guard
+            let sDangling = s { change = splitCoin extraChng (change s) }
+            let Fee φ_dangling = estimateFee (feeEstimator opts) sDangling
+            -- We have `δ_dangling = φ_original` by construction of sDangling.
+            --
+            -- Proof:
+            --
+            -- δ_dangling = Σi_dangling - (Σo_dangling + Σc_dangling)
+            --            = Σi_original - (Σo_original + Σc_original + extraChng)
+            --            = Σi_original - (Σo_original + Σc_original) - extraChng
+            --            = δ_original - extraChng
+            --            = δ_original - (δ_original - φ_original)
+            --            = φ_original
+            let δ_dangling = φ_original
+            case φ_dangling `C.sub` δ_dangling of
+                -- we've left too much, but adding a change output would be more
+                -- expensive than not having it. Here we have two choices:
+                --
+                -- a) If the node allows unbalanced transaction, we can stop here
+                --    and do nothing. We'll leave slightly more than what's needed
+                --    for fees, but having an extra change output isn't worth it
+                --    anyway.
+                --
+                -- b) If we __must__ balance the transaction, then we can choose
+                --    to pay the extra cost by adding the change output and
+                --    continue trying to balance the transaction (likely, by
+                --    selecting another input).
+                Just remainder | φ_dangling >= δ_original ->
+                    case onDanglingChange opts of
+                        SaveMoney ->
+                            pure (s, Fee C.zero)
+                        PayAndBalance ->
+                            pure (sDangling, Fee remainder)
+
+                -- If however, adding the dangling change doesn't cost more than
+                -- not having it, we might as well add it to get the money and
+                -- continue balancing!
+                _otherwise ->
+                    reduceChangeOutputs opts sDangling
+
+        -- The only way to end-up here is if the user has provided an invalid
+        -- selection where outputs are trying to spend more than inputs. This is
+        -- simply forbidden.
+        _Nothing ->
```

Comment:
```suggestion
        Nothing ->
```

Reply:
Actually, we need a wildcard here because of the guards and preconditions on other branch. GHC is not capable of identifying that the conditions above are mutually exclusive and that the only remaining branch actually is `Nothing`.

Reply:
> Actually, we need a wildcard here because of the guards and preconditions on other branch. GHC is not capable of identifying that the conditions above are mutually exclusive and that the only remaining branch actually is `Nothing`.

Fair enough!